### ISSUE TITLE
Object passed to InputObject parameter is modified by module

### ISSUE THREAD

**Issue by RobDesideri** [RamblingCookieMonster/PSDepend#35]
I don't know if it is a feature or a bug.

## Issue

```
$myObj = @{
  PSDependOptions = @{
    Target = 'CurrentUser'
   }

   AwesomeModule  = @{
    Version= 'latest'
    Tags = 'deploy'
   }

   AnotherAwesomeModule  = @{
    Version= 'latest'
    Tags = 'test'
   }
}

Invoke-PSDepend -InputObject $myObj -Install -Import -Tags Deploy
```

After this command, `PSDependOptions` item is removed from `myObj`:

```
$myObj.Keys  | %  {$_}

AwesomeModule
AnotherAwesomeModule 
```

So, if i recall Invoke-PSDepend:

```
Invoke-PSDepend -InputObject $myObj -Install -Import -Tags Test
```

the module will be installed in `AllUser` , not in the `CurrentUser` (the setted target).

## Causes
I found the `Invoke-PSDepend.ps1` script pass `InputObject` by reference to internal `GetPSDependParams` variable [[code](https://github.com/RamblingCookieMonster/PSDepend/blob/master/PSDepend/Public/Invoke-PSDepend.ps1#L193)].
So, when `Parse-Dependency` function in `Get-Dependency.ps1` script remove the `PSDependOptions` item [[code](https://github.com/RamblingCookieMonster/PSDepend/blob/master/PSDepend/Public/Get-Dependency.ps1#L220)] also it is removed from original object.

## Solution
Clone the object instead of pass by reference in [this](https://github.com/RamblingCookieMonster/PSDepend/blob/master/PSDepend/Public/Invoke-PSDepend.ps1#L193) row:

`$GetPSDependParams.add('InputObject',$InputObject)` 

to

`$GetPSDependParams.add('InputObject', $($InputObject.Clone()))`

## Workaround
Clone the object before passing it to the parameter.

**Comment by RamblingCookieMonster** [comment]
good catch!  feel free to submit a pr, otherwise, will try to grab this over the weekend!

**Comment by RamblingCookieMonster** [comment]
Hiyo!

Sorry for the delay - just realized, might need a little extra:

* `InputObject` is an array, so will need to consider this (e.g. unroll and clone each into new array)
* Dependencies can have nested hashtables, requiring a deep clone

My plates a bit full, but if anyone wants to take this on it would be awesome : )

Cheers!
